Add new utils for descriptors

madeleyneVaca opened the pull request:
If an EC configures the flexspi during application code, memory mapped reads like are done in get_active_slot and get_app_at_slot may not return the proper data. This commit adds utils to get the necessary info to be able to read the right section of memory into a buffer(s) (not memory mapped; i.e. something like a flash driver) and then construct the descriptors directly from the buffer(s).

madeleyneVaca committed:
Add new utils for descriptors

If an EC configures the flexspi during application code, memory mapped reads like are done in get_active_slot and get_app_at_slot may not return the proper data. This commit adds utils to get the app slot number and the app descriptors base address. Application code can read the descriptor region into a buffer and then use AppImageDescriptor::from_region with that buffer and the slot number

JamesHuard commented on src/lib.rs:
```diff
@@ -188,6 +188,14 @@ impl BootableRegionDescriptors {
         .unwrap()
     }
 
+    pub fn get_app_descriptor_base_address(&self) -> u32 {
```
above, if we call from_address on this type, it will validate the app image descriptors using the memory mapped interface. Do we need to make an equivalent function, from_region or some such, that does that validation over a provided memory buffer instead?

madeleyneVaca replied:
good catch!

madeleyneVaca committed:
add two more util functions to load the descriptors

Adds from_region to allow a user to build the boot descriptors from a provided buffer instead of utilizing memory-mapped reads at an address. This is useful in case application code has already configured the flexspi module as not all configs will read bits in the same was as the FCB. Because of the pointer arithmetic in locating the app descriptors, this function is considered unsafe as special care MUST be given to the buffer argument to avoid invalid memory access.
Adds from_header_and_app_regions to allow a user to build the boot descriptors from two small buffers containing the bytes for the header and the bytes for the app descriptors separately. This is useful in case these chunks are not contiguous in memory.

JamesHuard reviewed (dismissed):
think you've got some clippy warnings around documentation
